File: apps/agent/creature_tiers.py
"""Creature tier player-level bands from the game mechanics bestiary."""

TIER_LEVEL_RANGES = {1: (1, 4), 2: (5, 8), 3: (9, 14), 4: (15, 20)}


def player_level_range_for_tier(tier: int) -> tuple[int, int]:
    if tier not in TIER_LEVEL_RANGES:
        raise ValueError(f"unknown creature tier {tier}")
    return TIER_LEVEL_RANGES[tier]


def tier_for_player_level(level: int) -> int:
    matches = [tier for tier, (low, high) in TIER_LEVEL_RANGES.items() if low <= level <= high]
    if len(matches) != 1:
        raise ValueError(f"player level {level} maps to {len(matches)} creature tiers")
    return matches[0]

File: packages/shared/src/entities/creature_tiers.ts
export const TIER_LEVEL_RANGES: Record<number, readonly [number, number]> = {
  1: [1, 4],
  2: [5, 8],
  3: [9, 14],
  4: [15, 20],
};

export function playerLevelRangeForTier(tier: number): readonly [number, number] {
  const band = TIER_LEVEL_RANGES[tier];
  if (!band) throw new Error(`unknown creature tier ${tier}`);
  return band;
}

export function tierForPlayerLevel(level: number): number {
  const matches = Object.entries(TIER_LEVEL_RANGES)
    .filter(([, [low, high]]) => Number.isInteger(level) && low <= level && level <= high)
    .map(([tier]) => Number(tier));
  if (matches.length !== 1)
    throw new Error(`player level ${level} maps to ${matches.length} creature tiers`);
  return matches[0]!;
}
